sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala

```diff
@@ -220,6 +220,23 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
       }
     }
   }
+
+  test("SPARK-40885: V1 write uses the sort with partitionBy operator") {
+    withTempPath { f =>
+      Seq((20, 30, "partition"), (15, 20, "partition"),
+        (30, 70, "partition"), (18, 40, "partition"))
+        .toDF("id", "sort_col", "p")
+        .repartition(1)
+        .sortWithinPartitions("p", "sort_col")
+        .write
+        .partitionBy("p")
```

Comment:
I'm not very sure what the problem is. `partitionBy("p")` requires sorting the data by `p` per partition. It should already be satisfied and we don't need to add it.

Reply:
Oh I probably get what is wrong. The `V1WriteCommand.requiredOrdering` is for partition-wise ordering, but `LogicalPlan.outputOrdering` is global ordering.

Reply:
> `partitionBy("p")` requires sorting the data by `p` per partition.

Why does `write.partitionBy("p")` require sorting the data by `p` per partition? I understand why `df.repartition($"p")` requires in-partition order by `p` (`GroupedIterator`). But `write.partitionBy("p")`?

Reply:
> I'm not very sure what the problem is.

See https://github.com/apache/spark/pull/38356#issuecomment-1302534716:

The `V1Writes` rule introduced in Spark 3.4 adds the `empty2null` to all nullable string partition columns. The modified `V1WriteCommand` then has a modified `write.requiredOrdering` but the old `write.query.outputOrdering`, which then do not match any more. In `FileFormatWriter`, the outer sort will be added because of that mismatch.

Reply:
Seems `Project` shouldn't extend `OrderPreservingUnaryNode`, as it needs to rewrite the child output ordering based on the project list. cc @ulysses-you @wangyum

Reply:
yeah we can rewrite the ordering to preverse the attribute before alias like `AliasAwareOutputExpression`. But empty2null is special, it is not just an alias. Since we pull out the empty2null, we should tell project the ordering of empty2null(p) is same with p. One idea may be we should not pull out empty2null and leave it at original place..

Reply:
Be aware that `empty2null(p)` does not preserve **any** ordering of `p`, e.g. it does not preserve `NULL LAST` or ordering by `p` and `x`.

Reply:
we only add empty2null for string type partition columns, maybe we shouldn't optimize this case anyway.

Reply:
Adding `empty2null` throws null values and empty values into same partition, and the user has no way to make Spark treat them as distinct values. But changing this smells like a breaking change, unless some config allows to bring back the current behaviour.